cloudinit/sources/DataSourceScaleway.py

```diff
@@ -199,15 +209,15 @@ def __init__(self, sys_cfg, distro, paths):
 
     def _crawl_metadata(self):
         resp = url_helper.readurl(
-            self.metadata_address, timeout=self.timeout, retries=self.retries
+            self.metadata_address, headers=self.headers, headers_redact=self.headers_redact, timeout=self.timeout, retries=self.retries
         )
         self.metadata = json.loads(util.decode_binary(resp.contents))
 
         self.userdata_raw = query_data_api(
-            "user-data", self.userdata_address, self.retries, self.timeout
+            "user-data", self.userdata_address, self.headers, self.headers_redact, self.retries, self.timeout
```

Comment:
When cloud-init runs for the first time, it creates a pickle copy of the objects created in the first run including the **DatasourceScaleway** class. If the pickle copy exists, the existing data is used to initialize the classes.
So when this runs for the second time the **__init__()** method picks up the existing **DatasourceScaleway** class and does not create a new one (i.e. after an upgrade of the cloud-init package). This means that the self.headers and self.headers_redact attributes do not exist in the class and that these assignment fail with the followin backtrace : 
```
------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 761, in status_wrapper
    ret = functor(name, args)
  File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 433, in main_init
    init.apply_network_config(bring_up=bring_up_interfaces)
  File "/usr/lib/python3/dist-packages/cloudinit/stages.py", line 894, in apply_network_config
    and not event_enabled_and_metadata_updated(EventType.BOOT_LEGACY)
  File "/usr/lib/python3/dist-packages/cloudinit/stages.py", line 881, in event_enabled_and_metadata_updated
    and self.datasource.update_metadata_if_supported([event_type])
  File "/usr/lib/python3/dist-packages/cloudinit/sources/__init__.py", line 796, in update_metadata_if_supported
    result = self.get_data()
  File "/usr/lib/python3/dist-packages/cloudinit/sources/__init__.py", line 355, in get_data
    return_value = self._get_data()
  File "/usr/lib/python3/dist-packages/cloudinit/sources/DataSourceScaleway.py", line 231, in _get_data
    util.log_time(
  File "/usr/lib/python3/dist-packages/cloudinit/util.py", line 2606, in log_time
    ret = func(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/cloudinit/sources/DataSourceScaleway.py", line 212, in _crawl_metadata
    self.metadata_address, headers=self.headers, headers_redact=self.headers_redact, timeout=self.timeout, retries=self.retries
AttributeError: 'DataSourceScaleway' object has no attribute 'headers'
```
The modification needs to take into account that self.headers & self.headers_redact may not exist.

Reply:
The simplest is to make sure that those attributes exist when you get in the method : 
```
    def _crawl_metadata(self):
        # Stay backward compatible with classes w/o these attributes
        self.headers = getattr(self, 'headers', None)
        self.headers_redact = getattr(self, 'headers_redact', None)
```

Reply:
It's updated, it continue to works :)